README.md
# MMM-Canteen

**MMM-Canteen** is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror) project.

It shows the the menu including the prices of canteens from universities in Germany and Switzerland (based on [openmensa.org](https://openmensa.org)).

## Screenshot

![Example of a canteen in Jena](img/example1.png)

## Installation

Just clone the module into your modules folder of your MagicMirror² and execute `npm install --omit=dev` in the module’s directory:

```bash
git clone https://github.com/k-0/MMM-Canteen
cd MMM-Canteen
npm install --omit=dev
```

## Update

Go to the module’s folder inside MagicMirror modules folder and pull the latest version from GitHub and install:

```bash
git pull
npm install --omit=dev
```

## Using the module

### Configuration

To use this module, add it to the `config.js` file. Here is an example:

```javascript
{
    module: 'MMM-Canteen',
    position: 'bottom_center',
    config: {
        canteenName: 'Mensa am Park',
        canteen: 63,
        status: "employees",
        switchTime: "16:00"
    }
}
```

### Configuration options

<!-- prettier-ignore-start -->
| Option           | Description                   | Type    | Default                                  |
| ---------------- | ----------------------------- | ------- | ---------------------------------------- |
| `updateInterval` | Interval to update data       | Integer | `600000` (= 10 minutes)                  |
| `canteen`        | ID from the openmensa.org url | Integer | `63` (= Mensa am Park, Uni Leipzig)      |
| `status`         | Your status  <br> **Possible values:** `"employees"`, `"students"`, `"pupils"`, `"others"` | String | `"employees"` |
| `truncate`       | Truncate more than x letters  | Integer | `100`                                    |
| `debug`          | Debugging                     | Boolean | `false`                                  |
| `canteenName`    | Name of the canteen           | String  | `"Mensa am Park"`                        |
| `switchTime`     | Shows the menu from next day, if switchTime < now | Timestamp (HH:mm) | `"16:00"`  |
| `animationSpeed` | Speed of the update animation (in milliseconds).<br>If you don't want that the module blinks during an update, set the value to `0`. <br> **Possible values:** `0` - `5000` | Integer | `500` |
<!-- prettier-ignore-end -->

## Contributing

If you find any problems, bugs or have questions, please [open a GitHub issue](https://github.com/k-0/MMM-Canteen/issues) in this repository.

Pull requests are of course also very welcome 🙂


CHANGELOG.md
# MMM-Canteen Change Log

## [2.3.0] - 2023-08-15

- Remove 'node-fetch' dependency (use build-in fetch API instead)

Details: <https://github.com/KristjanESPERANTO/MMM-Canteen/compare/v.2.2.0...v.2.3.0>

## [2.2.0] - 2022-12-21

- Add veggie column 🌱

Details: <https://github.com/KristjanESPERANTO/MMM-Canteen/compare/v.2.1.0...v.2.2.0>

## [2.1.0] - 2022-12-20

- Add support for multiple instances 🙂

Details: <https://github.com/KristjanESPERANTO/MMM-Canteen/compare/v.2.0.0...v.2.1.0>

## [2.0.0] - 2022-12-18

First fork version 🚀

- Handle closed days: Inspired by the fork of joshua-martius: <https://github.com/joshua-martius/MMM-Canteen>
- Replace deprecated 'request' by 'node-fetch'
- Some rework like linting and formatting
- Optimize CSS
- Optimze installation process
- Fix issues

Details: <https://github.com/KristjanESPERANTO/MMM-Canteen/compare/v.1.1.0...v.2.0.0>


node_helper.js
const NodeHelper = require("node_helper");
const moment = require("moment");
const Log = require("logger");

const fetch = (...args) =>
  // eslint-disable-next-line no-shadow
  import("node-fetch").then(({ default: fetch }) => fetch(...args));

module.exports = NodeHelper.create({
  start() {
    Log.log(`Starting module helper: ${this.name}`);
  },

  socketNotificationReceived(notification, payload) {
    if (notification === "CONFIG") {
      this.config = payload;
      this.collectData();
      const self = this;
      setInterval(() => {
        self.collectData();
      }, 60 * 1000);
    }
  },

  async collectData() {
    let today;
    if (moment() < moment(this.config.switchTime, "HH:mm")) {
      today = moment().add(1, "days").format("YYYY-MM-DD");
    } else {
      today = moment().add(1, "days").format("YYYY-MM-DD");
    }
    const requestURL = `https://openmensa.org/api/v2/canteens/${this.config.canteen}/days/${today}/meals`;
    // Log.log(requestURL);
    const self = this;

    try {
      const response = await fetch(requestURL);

      if (response.status === 404) {
        Log.info("Kantine hat heut dicht!");
        self.sendSocketNotification("CLOSED", null);
      } else {
        const data = await response.json();
        // Log.log(data);
        self.sendSocketNotification("MEALS", data);
      }
    } catch (error) {
      Log.error(error);
    }
  }
});


MMM-Canteen.js
/* Magic Mirror
 * Module: MMM-Canteen
 */
Module.register("MMM-Canteen", {
  defaults: {
    updateInterval: 10 * 60 * 1000, // 10 minutes
    canteen: 838,
    status: "employees", // choose between ["employees", "students", "pupils", "others"]
    truncate: 100,
    switchTime: "16:00",
    debug: false,
    canteenName: "Kantine",
    animationSpeed: 500
  },

  loading: true,
  closed: false,
  meals: [],

  start() {
    Log.info(`Starting module: ${this.name}`);
    this.sendSocketNotification("CONFIG", this.config);
  },

  getStyles() {
    return ["MMM-Canteen.css"];
  },

  getTemplate() {
    return "MMM-Canteen.njk";
  },

  getTemplateData() {
    this.log("Updating template data");
    return {
      today:
        moment() < moment(this.config.switchTime, "HH:mm")
          ? moment().format("DD.MM.YYYY")
          : moment().add(1, "days").format("DD.MM.YYYY"),
      config: this.config,
      loading: this.loading,
      meals: this.meals,
      closed: this.closed
    };
  },

  socketNotificationReceived(notification, payload) {
    Log.info(`Socket Notification received: ${notification}`);
    this.loading = false;
    if (notification === "MEALS") {
      if (payload.length) {
        this.closed = false;
        this.meals = payload;
        this.log(this.meals);
      }
    } else if (notification === "CLOSED") {
      this.log("Mensa hat heute geschlossen!");
      this.closed = true;
    }
    this.updateDom(this.config.animationSpeed);
  },

  log(msg) {
    if (this.config && this.config.debug) {
      Log.info(`${this.name}: ${JSON.stringify(msg)}`);
    }
  }
});
